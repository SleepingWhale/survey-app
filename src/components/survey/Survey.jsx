import React, { PureComponent, Fragment } from 'react';
import { ControlSwitch } from '../controls';
import { SubmitViewContainer } from '../submitView';
import { Validators, validationMessages, types } from '../../common';
import { UserInfo } from '../UserInfor';

export class Survey extends PureComponent {
  state = {
    isValid: true,
    isSubmitted: false
  };

  validate = () => {
    const { question } = this.props;

    if (!Object.hasOwnProperty.call(types, question.type)) {
      return true;
    }

    return Validators[question.type](question.reply);
  };

  handleSubmit = () => {
    const isValid = this.validate();

    if (isValid) {
      this.setState({ isValid: true, isSubmitted: true });
    } else {
      this.setState({ isValid: false });
    }
  };

  handleNext = () => {
    const { onClickNext } = this.props;
    const isValid = this.validate();

    if (isValid) {
      this.setState({ isValid: true });
      onClickNext();
    } else {
      this.setState({ isValid: false });
    }
  };

  handleChange = value => {
    const { onChange } = this.props;
    onChange(value);
  };

  render() {
    const { question, quantity, questionIndex, onClickPrevious } = this.props;
    const { isValid, isSubmitted } = this.state;
    const previousDisabled = questionIndex === 0;
    const isLastQuestion = quantity === questionIndex + 1;
    const { reply, text, id, type } = question;

    return (
      <div className="container py-3">
        <div className="row">
          <div className="col-md-6 mx-auto">
            <div className="card border-primary">
              <div
                className="card-header bg-primary text-white"
                id="survey-header"
              >
                <div className="d-flex justify-content-between align-items-center">
                  <div>
                    {isSubmitted
                      ? 'Submitted'
                      : `${questionIndex + 1} / ${quantity}`}
                  </div>
                  <UserInfo />
                </div>
              </div>
              <div className="card-body">
                {isSubmitted ? (
                  <div className="card-text">
                    <SubmitViewContainer />
                  </div>
                ) : (
                  <Fragment>
                    <ControlSwitch
                      onChange={this.handleChange}
                      value={reply}
                      question={text}
                      id={id}
                      type={type}
                      key={id}
                    />
                    {!isValid && (
                      <div className="form-group" id="survey-validation">
                        <span className="badge badge-danger">
                          {' '}
                          {validationMessages[type]}{' '}
                        </span>
                      </div>
                    )}
                    <div className="d-flex justify-content-between">
                      <button
                        type="button"
                        className="btn btn-outline-primary"
                        onClick={onClickPrevious}
                        disabled={previousDisabled}
                        id="survey-button-previous"
                      >
                        Previous
                      </button>
                      {isLastQuestion ? (
                        <button
                          type="button"
                          className="btn btn-outline-primary"
                          onClick={this.handleSubmit}
                          id="survey-button-submit"
                        >
                          Submit
                        </button>
                      ) : (
                        <button
                          type="button"
                          className="btn btn-outline-primary"
                          onClick={this.handleNext}
                          id="survey-button-next"
                        >
                          Next
                        </button>
                      )}
                    </div>
                  </Fragment>
                )}
              </div>
            </div>
          </div>
        </div>
      </div>
    );
  }
}
