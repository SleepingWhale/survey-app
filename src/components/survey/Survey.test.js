import React from 'react';
import { shallow } from 'enzyme';
import { Survey } from './Survey';

const question = {
  id: 'A01',
  text: 'What happened to your product?',
  reply: 'Got lost',
  type: 'string',
  next: 'A02'
};

describe('Survey', () => {
  it('should render correctly', () => {
    const onClickPreviousHandler = jest.fn();
    const onClickNextHandler = jest.fn();
    const component = shallow(
      <Survey
        question={question}
        quantity={3}
        questionIndex={0}
        onClickPrevious={onClickPreviousHandler}
        onClickNext={onClickNextHandler}
      />
    );
    expect(component).toMatchSnapshot();
  });

  it('should render correct header', () => {
    const onClickPreviousHandler = jest.fn();
    const onClickNextHandler = jest.fn();
    const component = shallow(
      <Survey
        question={question}
        quantity={3}
        questionIndex={2}
        onClickPrevious={onClickPreviousHandler}
        onClickNext={onClickNextHandler}
      />
    );

    expect(component.find('#survey-header').contains('3 / 3')).toBeTruthy();
    component.find('#survey-button-submit').simulate('click');
    expect(component.find('#survey-header').contains('Submitted')).toBeTruthy();
  });

  it('should leaf questions', () => {
    const onClickPreviousHandler = jest.fn();
    const onClickNextHandler = jest.fn();
    const component = shallow(
      <Survey
        question={question}
        quantity={3}
        questionIndex={1}
        onClickPrevious={onClickPreviousHandler}
        onClickNext={onClickNextHandler}
      />
    );

    component.find('#survey-button-next').simulate('click');
    expect(onClickNextHandler).toHaveBeenNthCalledWith(1);

    component.find('#survey-button-previous').simulate('click');
    expect(onClickNextHandler).toHaveBeenNthCalledWith(1);
  });

  it('should not leaf if value is invalid', () => {
    const onClickPreviousHandler = jest.fn();
    const onClickNextHandler = jest.fn();
    const component = shallow(
      <Survey
        question={{
          id: 'A01',
          text: 'What happened to your product?',
          reply: 'text',
          type: 'number',
          next: 'A02'
        }}
        quantity={3}
        questionIndex={1}
        onClickPrevious={onClickPreviousHandler}
        onClickNext={onClickNextHandler}
      />
    );

    component.find('#survey-button-next').simulate('click');
    expect(onClickNextHandler).not.toHaveBeenCalled();
    expect(component.exists('#survey-validation')).toBeTruthy();
  });

  it('should not leaf back when on first question', () => {
    const onClickPreviousHandler = jest.fn();
    const onClickNextHandler = jest.fn();
    const component = shallow(
      <Survey
        question={question}
        quantity={3}
        questionIndex={0}
        onClickPrevious={onClickPreviousHandler}
        onClickNext={onClickNextHandler}
      />
    );

    component.find('#survey-button-previous').simulate('click');
    expect(onClickNextHandler).not.toHaveBeenCalled();
  });
});
