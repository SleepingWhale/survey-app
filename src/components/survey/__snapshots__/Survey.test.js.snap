// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Survey should render correctly 1`] = `
<div
  className="container py-3"
>
  <div
    className="row"
  >
    <div
      className="col-md-6 mx-auto"
    >
      <div
        className="card border-primary"
      >
        <div
          className="card-header bg-primary text-white"
          id="survey-header"
        >
          1 / 3
        </div>
        <div
          className="card-body"
        >
          <ControlSwitch
            id="A01"
            key="A01"
            onChange={[Function]}
            question="What happened to your product?"
            type="string"
            value="Got lost"
          />
          <div
            className="d-flex justify-content-between"
          >
            <button
              className="btn btn-outline-primary"
              disabled={true}
              id="survey-button-previous"
              onClick={[MockFunction]}
              type="button"
            >
              Previous
            </button>
            <button
              className="btn btn-outline-primary"
              id="survey-button-next"
              onClick={[Function]}
              type="button"
            >
              Next
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
